src/main/java/sc/fiji/bdvpg/scijava/services/SourceAndConverterService.java

```diff
@@ -226,6 +240,27 @@ public void registerPopupSourcesAction(Consumer<SourceAndConverter[]> action, St
         }
     }
 
+    @Override
+    public SourceAndConverter getSourceAndConverterFromSource( Source source )
+    {
+        final List< SourceAndConverter > sacs = getSourceAndConverters();
+
+        for ( SourceAndConverter sac : sacs )
+            if ( sac.getSpimSource().equals( source ) )
+                return sac;
+
+        // Try again, with unwrapping; (TODO: do we need this?)
+        if ( source instanceof TransformedSource )
+            source = ( ( TransformedSource ) source ).getWrappedSource();
+
+
+        for ( SourceAndConverter sac : sacs )
+            if ( sac.getSpimSource().equals( source ) )
+                return sac;
+
+        return null;
+    }
```

Comment:
This part I have major concerns. First part :
* a Source can have muliply linked `SourceAndConverter`
* a TransformedSource is different from a Source, so we should be able to set different projectors and properties if we deal with a Source or one of its transformations

Reply:
Yes, I had the same problem. I wrote a some text about it below, which you probably did not see. I paste it here again:

---

The AccumulateProjectorFactory in bdv work on a list of Source and not SourceAndConverter, which is quite unfortunate for us, because I need to get the metadata (projectionMode) for each Source. I thus added a function to our SourceAndConverterService that returns the SourceAndConverter for a Source. I am now very confused, because in our framework (e.g. due to SourceAndConverterDuplicator) there could be multiple SourceAndConverter for the same Source java object, in principle with different metadata. Currently the function I added (i.e., getSourceAndConverterFromSource) simply returns the first matching SourceAndConverter.

I am not sure how to handle this properly, i.e. how to figure out to which SourceAndConverter the Source belongs.

Do you have some ideas?

Question that come to my mind is:

Do we really need the SourceAndConverterDuplicator or can we avoid having this? I checked for its usages and it is only used by the SourcesDuplicatorCommand. As the code does not need it, do your users need it or could you provide similar functionality differently?
If we really need the SourceAndConverterDuplicator, could we wrap the Source that we put into the new SourceAndConverter such that it becomes a different and unique java object? E.g., make a DuplicatedSource( source )?
What are the other places where we wrap a Source into a SourceAndConverter? Could we also there wrap the Source into a unique object?
Could/Should we enforce that the metadata for a Source is unique? That is, we would not allow different SourceAndConverter for the same Source to have different metadata?
If you do not know a solution I would ask on gitter.

Reply:
Here's how I dealt with this issue in the remove from display :

https://github.com/bigdataviewer/bigdataviewer-playground/blob/c4f9b34607811e23180b01a8d6eee279cedff522/src/main/java/sc/fiji/bdvpg/scijava/services/BdvSourceAndConverterDisplayService.java#L249-L256